apps/bench.py

```diff
@@ -0,0 +1,1206 @@
+            self.bytes_sent += len(packet)
+            await self.packet_io.send_packet(packet)
+
+        await self.done.wait()
+        print(color('=== Done!', 'magenta'))
+
+    def on_packet_received(self, packet):
+        try:
+            packet_type, _ = parse_packet(packet)
+        except ValueError:
+            return
+
+        if packet_type == PacketType.ACK:
+            elapsed = time.time() - self.start_time
+            average_tx_speed = self.bytes_sent / elapsed
+            print(
+                color(
+                    f'@@@ Received ACK. Speed: average={average_tx_speed:.4f}'
+                    f' ({self.bytes_sent} bytes in {elapsed:.2f} seconds)',
+                    'green',
+                )
+            )
+            self.done.set()
+
+
+# -----------------------------------------------------------------------------
+# Receiver
+# -----------------------------------------------------------------------------
+class Receiver:
+    def __init__(self, packet_io):
+        self.reset()
+        self.packet_io = packet_io
+        self.packet_io.packet_listener = self
+        self.done = asyncio.Event()
+
+    def reset(self):
+        self.expected_packet_index = 0
+        self.start_timestamp = 0.0
+        self.last_timestamp = 0.0
+        self.bytes_received = 0
+
+    def on_packet_received(self, packet):
+        try:
+            packet_type, packet_data = parse_packet(packet)
+        except ValueError:
+            return
+
+        now = time.time()
+
+        if packet_type == PacketType.RESET:
+            print(color('=== Received RESET', 'magenta'))
+            self.reset()
+            self.start_timestamp = now
+            return
+
+        try:
+            packet_flags, packet_index = parse_packet_sequence(packet_data)
+        except ValueError:
+            return
+        print(
+            f'<<< Received packet {packet_index}: '
+            f'flags=0x{packet_flags:02X}, {len(packet)} bytes'
+        )
+
+        if packet_index != self.expected_packet_index:
+            print(
+                color(
+                    f'!!! Unexpected packet, expected {self.expected_packet_index} '
+                    f'but received {packet_index}'
+                )
+            )
+
+        elapsed_since_start = now - self.start_timestamp
+        elapsed_since_last = now - self.last_timestamp
+        self.bytes_received += len(packet)
+        instant_rx_speed = len(packet) / elapsed_since_last
+        average_rx_speed = self.bytes_received / elapsed_since_start
+        print(
+            color(
+                f'Speed: instant={instant_rx_speed:.4f}, average={average_rx_speed:.4f}',
+                'yellow',
+            )
+        )
+
+        self.last_timestamp = now
+        self.expected_packet_index = packet_index + 1
+
+        if packet_flags & PACKET_FLAG_LAST:
+            asyncio.create_task(
```

Comment:
Because of https://github.com/python/cpython/issues/88831, I suggest to push the packet to a queue and to loop get/send in // of `self.done.wait()` in `async def run(self)` bellow

Reply:
Fixed in latest commit.

Reply:
I just thought about that but we call it spawn instead of run?

Reply:
Yes, good point, `spawn` is a better name for it.